File: 97_Interleaving_String/isInterleave.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
    public:
        bool isInterleave(string s1, string s2, string s3) {
            int m = s1.size(), n = s2.size();
            // T[i][j]: whether s3[:i+j] is the interleaving of s1[:i] and s2[:j]
            if((m + n) != s3.size()) return false;
            vector<vector<bool> > T(m + 1, vector<bool>(n + 1, false));
            T[0][0] = true;

            // s2 is NULL.
            for(int i = 1; i <=m; i++)
                if(s1[i - 1] == s3[i - 1])
                    T[i][0] = T[i - 1][0];

            // s1 is NULL.
            for(int j = 1; j <= n; j++)
                if(s2[j - 1] == s3[j - 1])
                    T[0][j] = T[0][j - 1];

            for(int i = 1; i <= m; i++) {
                for(int j = 1; j <= n; j++) {
                    if(s1[i - 1] == s3[i + j - 1] && s2[j - 1] != s3[i + j - 1])
                        T[i][j] = T[i - 1][j];
                    else if(s1[i - 1] != s3[i + j - 1] && s2[j - 1] == s3[i + j - 1])
                        T[i][j] = T[i][j - 1];
                    else if(s1[i - 1] == s3[i + j - 1] && s2[j - 1] == s3[i + j - 1])
                        T[i][j] = T[i - 1][j] || T[i][j - 1];
                }
            }

            return T[m][n];
        }
};

File: 97_Interleaving_String/isInterleave.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        if not s1 or  not s2 or  not s3:
            return True

        if not s3:
            return False

        return (s3[0] == s1[0] and self.isInterleave(s1[1:], s2, s3[1:])) \
                or (s3[0] == s2[0] and self.isInterleave(s1, s2[1:], s3[1:]))

sol = Solution()
s1 = "aabcc"
s2 = "dbbca"
s3 = "aadbbcbcac"

print sol.isInterleave(s1, s2, s3)

